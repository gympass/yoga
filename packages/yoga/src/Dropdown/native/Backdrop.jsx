import React, { useEffect, useState } from 'react';
import styled, { withTheme } from 'styled-components';
import { node, string, func, bool } from 'prop-types';
import {
  Animated,
  TouchableWithoutFeedback,
  Easing,
  Modal,
} from 'react-native';

import Text from '../../Text';

const ClosableArea = styled(Animated.View)`
  ${({
    theme: {
      yoga: { colors },
    },
  }) => `
    position: relative;
    width: 100%;
    height: 100%;

    background-color: ${colors.text.primary};
  `}
`;

const ContentWrapper = styled(Animated.View)`
  position: absolute;
  justify-content: flex-end;

  width: 100%;
`;

const Content = styled.View`
  ${({
    theme: {
      yoga: {
        components: { dropdown },
      },
    },
  }) => `
    justify-content: center;
    width: 100%;

    background-color: ${dropdown.backdrop.content.backgroundColor};
    border-top-left-radius: ${dropdown.backdrop.content.border.radius.topLeft}px;
    border-top-right-radius: ${dropdown.backdrop.content.border.radius.topRight}px;
  `}
`;

const Title = styled(Text.Bold)`
  ${({
    theme: {
      yoga: {
        components: { dropdown },
      },
    },
  }) => `
    width: 100%;
    padding:
      ${dropdown.backdrop.content.title.padding.top}px
      ${dropdown.backdrop.content.title.padding.right}px
      ${dropdown.backdrop.content.title.padding.bottom}px
      ${dropdown.backdrop.content.title.padding.left}px;

    font-size: ${dropdown.backdrop.content.title.font.size}px;

    text-align: center;
  `}
`;

const Backdrop = ({
  visible = false,
  title = null,
  children,
  onClose = () => {},
  theme: {
    yoga: {
      transition,
      components: { dropdown },
    },
  },
  ...props
}) => {
  const [isOpen, toggleIsOpen] = useState(visible);
  const [backgroundAnimation] = useState(new Animated.Value(0));
  const [contentAnimation] = useState(
    new Animated.Value(-dropdown.backdrop.content.height),
  );

  const animate = (animation, toValue) =>
    Animated.timing(animation, {
      toValue,
      duration: transition.duration[0],
      easing: Easing.out(Easing.ease),
      useNativeDriver: false,
    }).start(() => !visible && toggleIsOpen(false));

  useEffect(() => {
    if (visible) toggleIsOpen(true);
    animate(backgroundAnimation, visible ? 0.5 : 0);
    animate(contentAnimation, visible ? 0 : -dropdown.backdrop.content.height);
  }, [visible]);

  return (
    isOpen && (
      <Modal {...props} transparent animationType="none">
        <TouchableWithoutFeedback onPress={onClose}>
          <ClosableArea style={{ opacity: backgroundAnimation }} />
        </TouchableWithoutFeedback>
        <ContentWrapper style={{ bottom: contentAnimation }}>
          <Content>
            {title && <Title>{title}</Title>}
            {children}
          </Content>
        </ContentWrapper>
      </Modal>
    )
  );
};

Backdrop.propTypes = {
  visible: bool,
  title: string,
  children: node.isRequired,
  onClose: func,
};

export default withTheme(Backdrop);
