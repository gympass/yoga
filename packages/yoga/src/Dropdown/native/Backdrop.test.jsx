import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from '../..';
import Backdrop from './Backdrop';

jest.useFakeTimers();

describe('<Backdrop />', () => {
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      const { toJSON } = render(
        <ThemeProvider>
          <Backdrop title="Find an Activity">Some Content</Backdrop>
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });

    it('should match snapshot when visible', () => {
      const { toJSON } = render(
        <ThemeProvider>
          <Backdrop visible title="Find an Activity">
            Some Content
          </Backdrop>
        </ThemeProvider>,
      );

      expect(toJSON()).toMatchSnapshot();
    });
  });
});
