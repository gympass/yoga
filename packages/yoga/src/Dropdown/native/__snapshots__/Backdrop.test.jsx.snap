// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Backdrop /> Snapshots should match snapshot 1`] = `null`;

exports[`<Backdrop /> Snapshots should match snapshot when visible 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#231B22",
        "height": "100%",
        "opacity": 0,
        "position": "relative",
        "width": "100%",
      }
    }
  />
  <View
    collapsable={false}
    style={
      Object {
        "bottom": -320,
        "justifyContent": "flex-end",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderTopLeftRadius": 8,
            "borderTopRightRadius": 8,
            "justifyContent": "center",
            "width": "100%",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#231B22",
              "fontFamily": "Rubik",
              "fontSize": 16,
              "fontWeight": "700",
              "paddingBottom": 20,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 20,
              "textAlign": "center",
              "width": "100%",
            },
          ]
        }
      >
        Find an Activity
      </Text>
      Some Content
    </View>
  </View>
</Modal>
`;
